packages/vitest/src/runtime/mocker.ts

```diff
@@ -209,10 +209,21 @@ export class VitestMocker {
       const isModule = containerType === 'Module' || !!container.__esModule
       for (const { key: property, descriptor } of getAllMockableProperties(container)) {
         // Modules define their exports as getters. We want to process those.
-        if (!isModule) {
-          // TODO: Mock getters/setters somehow?
-          if (descriptor.get || descriptor.set)
-            continue
+        if (!isModule && descriptor.get) {
+          try {
+            const newDescriptor: PropertyDescriptor = {
+              enumerable: true,
+              configurable: true,
+              get: spyModule.fn(),
```

Comment:
Do we want to _mock_ getters/setters? I assumed we just place them on a new object as is, no? If not, that's different from jest, as I understand it, and needs to be documented in migration guide and mocking guide.

Reply:
> Do we want to _mock_ getters/setters? I assumed we just place them on a new object as is, no?

I think we want to mock them. Since they are functions it makes sense to treat them the same as any other function during the automocking process, which is to completely stub it out to ensure it doesn't cause any side effects. If you want to have partial implementation you can always do a custom `__mocks__` definition, use `importActual`, or just not mock it in the first place.

I can't think of a good reason to mock everything _except_ getters/setters. I see no reason why they should be special here, but if you have a compelling argument I'm happy to consider it.

> If not, that's different from jest, as I understand it, and needs to be documented in migration guide and mocking guide.

This is already a difference from Jest. Jest doesn't mock getters or setters _at all_, which is why I had implemented it the way I did in #1648.

Reply:
What do you mean by `Since they are functions`? If they are on first level, then the getter was added by Vite: 

```ts
const someValue = 3
Object.defineProperty(__vite_exports__, 'someValue', { get: () => someProperty })
```

> I can't think of a good reason to mock everything except getters/setters. I see no reason why they should be special here, but if you have a compelling argument I'm happy to consider it.

In the same message you are saying that Jest doesn't mock them. I see no reason to automock getters/setters at all, to be honest.

> Jest doesn't mock getters or setters at all, which is why I had implemented it the way I did in https://github.com/vitest-dev/vitest/pull/1648.

This doesn't explain why you just didn't put the same descriptor on new object. Why does a new object has a hole in it, compared to the original?

Reply:
> What do you mean by `Since they are functions`? If they are on first level, then the getter was added by Vite:
> 
> ```ts
> const someValue = 3
> Object.defineProperty(__vite_exports__, 'someValue', { get: () => someProperty })
> ```

Getters are functions with 0 arguments that are accessed without parenthesis. Even in your example right there the getter is defined as an arrow function that returns `someProperty`.

---

> In the same message you are saying that Jest doesn't mock them. I see no reason to automock getters/setters at all, to be honest.

This PR exists because of #1899. If we want to just stick with Jest compatibility then close this PR and mark #1899 as wontfix. 

---

> Also, your implementation doesn't allow restoring mocks to original setters/getters.

This is true, but imo that doesn't matter. Why would you want to restore the mocks on an automocked import anyway? Plus, if we're going with the "the way Jest does it" argument, Jest doesn't allow restoring _any_ automocked properties. If you want to mock a getter and restore it later just use `spyOn` yourself.

---

> This doesn't explain why you just didn't put the same descriptor on new object. Why does a new object has a hole in it, compared to the original?

What do you mean "has a hole in it"? This PR _fills_ a hole in the mocked objects.

Reply:
> Getters are functions with 0 arguments that are accessed without parenthesis

I know what a function is. I don't know however how this is important to this discussion? Why do we need to treat all "technically" functions the same?

> This PR exists because of https://github.com/vitest-dev/vitest/issues/1899. If we want to just stick with Jest compatibility then close this PR and mark https://github.com/vitest-dev/vitest/issues/1899 as wontfix, and I'll put up another PR removing this FIXME:

This PR fixes it in a way that is not conventional. I would prefer to just save the descriptors as they are, no need to "auto" mock it.

> Why would you want to restore the mocks on an automocked import anyway?

"Because I want to" is enough of a reason, when implementation is a one liner and doesn't contradict anything else here. By your logic of "it's a function, it should follow what other functions do", it should be restorable, since it's possible to restore automocked functions. But it doesn't matter, because we don't want to automock setters/getters, just move them on a new object.

> What do you mean "has a hole in it"? This PR fills a hole in the mocked objects.

I was refering to your previous PR that had a hole, because you were saying that you implemented in this way intentionally ("which is why I had implemented it the way I did").

Reply:
> Why do we need to treat all "technically" functions the same?

> I would prefer to just save the descriptors as they are, no need to "auto" mock it.

> But it doesn't matter, because we don't want to automock setters/getters, just move them on a new object.

Say you have the following class:

```js
import { readFile } from 'node:fs/promises';

class SettingsWrapper {
  constructor(settingsFilePath) {
    this.settingsFilePath = settingsFilePath;
  }

  get settings() {
    return readFile(this.settingsFilePath, { encoding: 'utf-8' }).then(
      (json) => {
        return JSON.parse(json);
      },
    );
  }
}
```

If we just blindly copy the descriptor over to the new object then accessing the `settings` getter on a mocked instance would perform an _actual disk read_. Sure you can argue that this is bad design, but it's still possible and thus needs to be accounted for in how we handle automocking. Since getters and setters are functions they can do _anything_ a normal function can do, which includes some _serious_ side effects that people want to mock out.

---

> I was referring to your previous PR that had a hole, because you were saying that you implemented in this way intentionally ("which is why I had implemented it the way I did").

I implemented it that way before because that's how Jest has been doing it for over 5 years<sup>1</sup>, and they seemed to be doing just fine.

--- 

This whole conversation has made me think we really have two options here: either do exactly what Jest does and don't mock getters/setters at all, and if you want to stub them out during tests just set them yourself<sup>2</sup>, or one-up Jest and automock the getters/setters as in this PR. I feel _very_ strongly that just blindly copying them over to the final object is the _wrong_ solution and _will_ cause problems in the future if we try to go that route.

---

<sup>1</sup> https://bambielli.com/til/2017-09-17-jest-doesnt-mock-getters/
<sup>2</sup> https://bambielli.com/til/2017-09-17-jest-doesnt-mock-getters/#option-2-just-assign-a-value

Reply:
As for the restorability of Mocks, that's *another* difference from Jest. Jest doesn't allow automatic restoring of _any_ automocked attributes. If you want the original again in Jest you _have_ to use `requireActual`. The only reason I _didn't_ mirror that behavior in my original PR is because some of the existing tests already relied on that behavior, and I was trying to make my changes have as few breakages as possible.

Reply:
> Say you have the following class

This is just a hypothetical situation. By your own logic, we can ask user to take care of it.

>  I feel very strongly that just blindly copying them over to the final object is the wrong solution and will cause problems in the future if we try to go that route.

This is how it was working before in Vitest. Just restore this behaviour. 

As I said before, getters/setters don't only exist on classes and objects. Our module system is implemented with getters/setters. Exported constants *SHOULD* stay the same. This is the end of the conversation.